crl/updater/batch.go

```diff
@@ -0,0 +1,256 @@
+package updater
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"sort"
+	"strings"
+	"time"
+
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/letsencrypt/boulder/crl"
+	"github.com/letsencrypt/boulder/issuance"
+)
+
+// RunOnce runs the entire update process once immediately. It processes each
+// configured issuer serially, and processes all of them even if an early one
+// encounters an error. All errors encountered are returned as a single combined
+// error at the end.
+func (cu *crlUpdater) RunOnce(ctx context.Context, atTime time.Time) (err error) {
+	defer func() {
+		// This func closes over the named return value `err`, so can reference it.
+		result := "success"
+		if err != nil {
+			result = "failed"
+		}
+		cu.tickHistogram.WithLabelValues("all", result).Observe(cu.clk.Since(atTime).Seconds())
+	}()
+	cu.log.Debugf("Ticking at time %s", atTime)
+
+	var errIssuers []string
+	for id := range cu.issuers {
+		// For now, process each issuer serially. This keeps the worker pool system
+		// simple, and processing all of the issuers in parallel likely wouldn't
+		// meaningfully speed up the overall process.
+		err := cu.updateIssuer(ctx, atTime, id)
+		if err != nil {
+			cu.log.AuditErrf(
+				"Generating CRLs for issuer failed: number=[%d] issuer=[%s] err=[%s]",
+				(*big.Int)(crl.Number(atTime)), cu.issuers[id].Subject.CommonName, err)
+			errIssuers = append(errIssuers, cu.issuers[id].Subject.CommonName)
+		}
+	}
+
+	if len(errIssuers) != 0 {
+		return fmt.Errorf("%d issuers failed: %v", len(errIssuers), strings.Join(errIssuers, ", "))
```

Comment:
I believe that @aarongable does this in `updateIssuer()` for testing purposes. Though I agree that it's probably much easier for an operator to read a message if the numbers of the failed shards have some kind of order to them.

Reply:
Same with this -- totally agreed that this can/should be better, but the code in batch.go is unchanged in this PR, so I'd prefer not mess with it here and now. We'll improve this in https://github.com/letsencrypt/boulder/issues/7066.